feat: Update PR template to something people might actually use (#243)
openedx/edx-cookiecutters: python-template/{{cookiecutter.placeholder_repo_name}}/.github/PULL_REQUEST_TEMPLATE.md

> **Description:** Describe in a couple of sentences what this PR adds

Reviewer: I'm not sure if there is a relevant OEP, or if we should just create an ADR in this repo, but the cookiecutter is making default decisions across repos, so it would be worth vetting changes like this across a wider audience.

Author: I've made an identical PR for event-bus-kafka so we can try it out there: https://github.com/openedx/event-bus-kafka/pull/29

Reviewer 2: I don't think an OEP or ADR is necessary because I think this change and its commit message are self-documenting. That said, it would be fair to circulate the PR around a few public Slack channels in order to expose it to a wider audience.

> **Description:** Describe in a couple of sentences what this PR adds

Reviewer 2: As an alternative to this PR: Why don't we delete this template from the cookiecutter, and instead add the PR template as an org-wide default to the `.github` repository? That way, updates to it will immediately apply to all repos without an explicit PR template, rather than only new repos.

Author: ...I had no idea you could have an org-wide PR template! I actually really like that, because 1) I am generally suspicious of the maintenance burden that results from each new file the cookiecutter produces, and 2) it still allows customization if a repo has special requirements. I may open a new PR for `.github`.

Reviewer 2: Nice. The only wrinkle is that not every repo has a changelog or versioning. But I'm sure you could figure out some creative wording to gloss over that.

Reviewer 2: Speaking of that, how do you want folks to deal with check boxes that just don't apply to their PR?

Remove them? Strikethrough? Check them anyway?

Author: Maybe the org-wide one should end with "(If any of these are _never_ applicable to this repo, please make a copy of this org-wide pull request template and trim it as appropriate.)" :laughing: It's pretty hard to have something both comprehensive _and_ relevant.

For inapplicable ones I personally tend to just check them off in the spirit of keeping things low-effort. I've suggested the language "Check off if complete or not applicable:" which hopefully would disambiguate that in practice.

Reviewer 2: > (If any of these are never applicable to this repo, please make a copy of this org-wide pull request template and trim it as appropriate.)

I like the sentiment but I fear that this would just take up space in the template and be ignored.

>  "Check off if complete or not applicable:" which hopefully would disambiguate that in practice.

How about "Delete if not applicable?" 

I think "I did this" versus "I don't think it's applicable" is a useful thing to communicate to reviewers. Particularly around documentation and manual testing.

Unless, of course, the purpose if this template is not to tell reviewers things, but is to remind authors to do things, in which case just checking the non-applicable boxes works fine.

Author: Yeah, more text is probably not the answer. :-P

I mostly think about it as reminding authors, although it's great for reviewers too. So I'd prefer ticking to deleting.

**Merge checklist:**

> Check off if complete or not applicable:

Reviewer 2: Sorry, I thought of one more thing 😬 

```suggestion
Check off if complete or not applicable:
- [ ] Backported to [latest and next named releases](https://docs.openedx.org/en/latest/community/release_notes/named_release_branches_and_tags.html?highlight=latest#latest-open-edx-release)
```

IMO that would remove the need for BTR to stick [this big warning at the top of the edx-platform PR template](https://github.com/openedx/edx-platform/blob/674bf2a040aa97fd2b98b8a513b0ec592fbeec8f/.github/pull_request_template.md?plain=1#L1-L20).

Author: Backports are usually only relevant for IDAs, right? And versioning and changelogs are usually only done for libraries. So maybe it makes sense to keep separate templates for the various cookiecutters after all.

Reviewer 2: It's a PITA so we try to avoid it, but we do need to backport fixes to libraries sometimes: https://openedx.atlassian.net/wiki/spaces/COMM/pages/2065367719/Making+a+pull+request+for+a+named+release#Backporting-to-a-library-repository

But anyway, to completely contradict my own suggestion: maybe this is trying to hard? I'm realizing that any truly exhaustive list of pre-merge items would be too long for a good PR template. I think there's virtue in having one simple, universal PR template, rather than trying to build a bunch of templates for different types of repos.

In terms of versioning and changelogs: Can we phrase it in a way that works for IDAs too? Maybe something like:

> - [Release process]() followed

Where "release process" is a link to a wiki page that explains or links out to:
* our backend library release process (version bump, changelog, whatever)
* our frontend library release process (version bump, GH release, whatever)
* the process for MFEs and IDAs, for:
  * core contributors (message relevant `#cc-*` channel)
  * 2U engineers (could link to an internal page)
* the backporting process & how to know whether it applies to your PR

I realize that building such a page would take some work...

Author: After more consideration, I think we actually want to go in the exact opposite direction: Customize each repo's PR template as much as possible. Some repos keep changelogs, others do not; some have versions, others are continuously released; for some, manual testing is appropriate, and for others it is not.... The more I tried to create a universal template, the larger it got.

Fundamentally, I think there are three critical principles in checklist design:

1. Make it as short as possible -- otherwise people get annoyed/fatigued/intimidated
2. Have zero irrelevant items for the vast majority of occasions -- otherwise people get in the habit of checking off items without considering them carefully
3. Have the easiest-to-forget things highest in the list, to increase the chance they are noticed

I don't think you can create a universal template and also stick to those items, particularly # 2. In fact, not having the template be too uniform probably hurts # 3, as it increases a sense of familiarity, which reduces the chance people will stop and think.

Anyway, take a look at the latest. The commit message describes the differences between the variations. If you hate it, I can back it out, but I think this approach has the best chance of increasing participation.

Reviewer 2: I agree with those checklist design principles. Once the checklists are designed, though, I think three things need to happen in order for them to be adopted:

1. **New repos get the correct PR template.** This is easy enough.
2. **Existing repositories are updated to use correct PR template.** (Yes, we could skip on updating existing repos. However, I don't see these checklists having an impact unless folks see them basically everywhere.)
   * For an org-wide template, it'd be easy. Just delete all existing PR templates 💥 
   * For customized templates, you'd need to go in and update every repo to the right PR template.
3. **Checklists are updated as repos & processes change over time.**
   * For an org-wide template, this would just mean editing the central PR template or supplementary wiki pages.
   * For customized templates, this would mean updating templates across repositories, every time things change. Either maintainers or a central team like Arch-BOM would need to take on this responsibility. (FWIW, tCRIL has [some new automation for cross-repo updates](https://github.com/openedx/terraform-github/blob/main/migrate/repo_checks.py). We've been using it to keep certain workflows consistent across repos. Still, merging all the generated PRs is annoying).

Personally, I think I would still choose an org-wide template just to avoid taking on the up-front cost of (2) and the maintenance burden of (3). That said, if customized templates seem worth the lift to you, then no opposition from me 👍🏻

Author: Yeah, I definitely feel the tension of "ugh, not another thing to keep updated", which I think I actually mentioned earlier. But PR templates are intimately tied to ownership, and different maintainers have different checklists (at least in their minds, if not explicitly written down.) It would actually be really cool if we could standardize maintainership, and then I think there'd be a really strong argument for standardized PR template checklists. But even then we'd have differences for libraries vs. IDAs, and GH doesn't have a way to say "I want Standard PR Template No. 5".

I'm OK with the idea of having to automate (or semi-automate) replacement of PR templates at some point. I think we're going to need to get better at that sort of thing eventually anyhow. One option would be to *fully* automate modification of PR templates that haven't been edited; that would be a pretty safe thing to do, since you could say "auto-merge any where the the diff looks *exactly* like this" and be confident that the automation wasn't screwing anything up. That's how I'd want to approach any such future change.

Reviewer 2: Sounds good 👍

Once you merge I recommend linking up with the #maintainership-pilot folks. They have a draft repo standards list going and it sounds like "appropriate PR template" ought to be on there.

- [ ] Version bumped
- [ ] Changelog record added
- [ ] Documentation updated (not only docstrings)

> - [ ] Commits are squashed

Reviewer 2: ```suggestion
- [ ] Commits are squashed into one or more [conventional commits](https://open-edx-proposals.readthedocs.io/en/latest/best-practices/oep-0051-bp-conventional-commits.html)
```

Author: Oh, Phil raised an interesting question in chat -- we already have a commitlint CI check. Still worth the extra text? Could go either way.

> - [ ] Noted any: Concerns, dependencies, migration issues, deadlines, tickets, testing instructions

Reviewer 2: ```suggestion
- [ ] Noted any: Concerns, dependencies, migration issues, deadlines, tickets
- Either:
  - [ ] Manually tested, with steps and/or screenshots provided
  - [ ] Manual testing is not necessary
```

At risk of bloating the list: *Did you test this, and if so, how?* is probably the number-one question I ask in PR reviews. I love that OpenCraft's PR template includes it as an item and I think it makes reviewing their PRs much easier. I suggest adapting it here.

Reviewer 2: I think those are not mutually exclusive. If a reviewer adds a user-facing feature, I'd want them to both add unit tests and check that it works manually.

Then, there are some changes where automatic testing isn't possible (eg CSS changes) and some changes where testing just isn't required (eg reformatting a file).
